lib/handlers/authorize-handler.js

```diff
@@ -0,0 +1,332 @@
+          var redirectUri = this.buildSuccessRedirectUri(uri, responseType);
+
+          this.updateResponse(response, redirectUri, state);
+
+          return code;
+        })
+        .catch(function(e) {
+          if (!(e instanceof OAuthError)) {
+            e = new ServerError(e);
+          }
+
+          var redirectUri = this.buildErrorRedirectUri(uri, e);
+
+          this.updateResponse(response, redirectUri, state);
+
+          throw e;
+        });
+    });
+};
+
+/**
+ * Generate authorization code.
+ */
+
+AuthorizeHandler.prototype.generateAuthorizationCode = function() {
+  if (this.model.generateAuthorizationCode) {
+    return Promise.try(this.model.generateAuthorizationCode);
+  }
+
+  return tokenUtil.generateRandomToken();
+};
+
+/**
+ * Get authorization code lifetime.
+ */
+
+AuthorizeHandler.prototype.getAuthorizationCodeLifetime = function() {
+  var expires = new Date();
+
+  expires.setSeconds(expires.getSeconds() + this.authorizationCodeLifetime);
+
+  return expires;
+};
+
+/**
+ * Get the client from the model.
+ */
+
+AuthorizeHandler.prototype.getClient = function(request) {
+  var clientId = request.body.client_id || request.query.client_id;
+
+  if (!clientId) {
+    throw new InvalidRequestError('Missing parameter: `client_id`');
+  }
+
+  if (!is.vschar(clientId)) {
+    throw new InvalidRequestError('Invalid parameter: `client_id`');
+  }
+
+  var redirectUri = request.body.redirect_uri || request.query.redirect_uri;
+
+  if (redirectUri && !is.uri(redirectUri)) {
+    throw new InvalidRequestError('Invalid request: `redirect_uri` is not a valid URI');
+  }
+
+  return Promise.try(this.model.getClient, clientId)
+    .then(function(client) {
+      if (!client) {
+        throw new InvalidClientError('Invalid client: client credentials are invalid');
+      }
+
+      if (!client.grants) {
+        throw new InvalidClientError('Invalid client: missing client `grants`');
+      }
+
+      if (!_.contains(client.grants, 'authorization_code')) {
+        throw new UnauthorizedClientError('Unauthorized client: `grant_type` is invalid');
+      }
+
+      if (!client.redirectUris || 0 === client.redirectUris.length) {
+        throw new InvalidClientError('Invalid client: missing client `redirectUri`');
+      }
+
+      if (redirectUri && !_.contains(client.redirectUris, redirectUri)) {
+        throw new InvalidClientError('Invalid client: `redirect_uri` does not match client value');
+      }
+
+      return client;
+    });
+};
+
+/**
+ * Get scope from the request.
+ */
+
+AuthorizeHandler.prototype.getScope = function(request) {
+  var scope = request.body.scope || request.query.scope;
+
+  if (!is.nqschar(scope)) {
+    throw new InvalidScopeError('Invalid parameter: `scope`');
+  }
+
+  return scope;
+};
+
+/**
+ * Get state from the request.
+ */
+
+AuthorizeHandler.prototype.getState = function(request) {
+  var state = request.body.state || request.query.state;
+
+  if (!state) {
```

Comment:
The concern with this breaking upgrade path is not that it's a breaking change or that it would require significant changes for users to upgrade. It's that you're forgetting out all the existing clients relying on the users using this module. If I have 100,000 applications built on my platform and I use node-oauth2-server and want to upgrade to the new version, then I'm left with a major problem for all of my users that currently don't pass a state for their applications. Since this is technically breaking the current spec, that is an issue of the module, not one to pass onto my users.